### ISSUE TITLE
Using EVP_bf_cbc() fails for encrypt, segfaults for decrypt

### ISSUE THREAD

**Issue by levaidaniel** [openssl/openssl#19634]
Recently upgraded to OpenSSL 3 and noticed that a specific use-case is broken in one of my apps. More specifically, when using blowfish crypto.

Attached a sample code as a text file while also including it inline at the end of the report to help reproduce.

It uses EVP_aes_256_cbc() for reference, just switch over `aes` with `bf` at both places (for encrypt and decrypt) to see the error:
```
        /* Turn on encryption */
        /*
        BIO_set_cipher(bio_chain, EVP_bf_cbc(), key, iv, 1);
        */
        BIO_set_cipher(bio_chain, EVP_aes_256_cbc(), key, iv, 1);
[...]
        /* Turn on decryption */
        /*
        BIO_set_cipher(bio_chain, EVP_bf_cbc(), key, iv, 1);
        */
        BIO_set_cipher(bio_chain, EVP_aes_256_cbc(), key, iv, 0);
```
Using AES, it runs fine:
```
$ clang -lcrypto -o test test.c && ./test
bio_chain: BIO_TYPE_CIPHER->BIO_TYPE_BASE64->BIO_TYPE_FILE
data(4) = data
wrote: 4
Encrypt OK
flushed bio_chain
pending write bytes from bio_chain: 0
pending bytes from bio_chain: 0
db_file size -> 25
read: 4
Decrypt OK
data(4) = data
```

Encryption fails, while decrypt segfaults with blowfish (cipher modes other than `cbc` are affected too):
```
$ clang -lcrypto -o test test.c && ./test
bio_chain: BIO_TYPE_CIPHER->BIO_TYPE_BASE64->BIO_TYPE_FILE
data(4) = data
BIO_write() error (don't retry)
ERROR: failed to encrypt
BIO_flush() error (don't retry)
db_file size -> 0
Segmentation fault (core dumped)
```

Doesn't seem to be the case with any other EVP_*().

I tested it on Arch and Fedora 36, Arch having OpenSSL 3.0.7, while Fedora is at 3.0.5, as of this writing.

```
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <openssl/bio.h>
#include <openssl/evp.h>


void            print_bio_chain(BIO *);


int
main(int argc, char *argv[])
{
        const char              *data = "data";
        char                    *read = NULL;

        const unsigned char     key[12] = { 'a', 's', 'd' };
        const unsigned char     key2[12] = { 'b', 's', 'd' };
        const unsigned char     iv[12] = { '1', '2', '3' };

        int     ret = 0, remaining = 0, pos = 0;

        BIO     *bio_chain = NULL;
        BIO     *bio_cipher = NULL;
        BIO     *bio_b64 = NULL;
        BIO     *bio_file = NULL;


        /* Setup */

        bio_file = BIO_new_file("./test.db", "w+");
        if (!bio_file) {
                printf("ERROR: BIO_new_file()");
                return(2);
        }
        BIO_set_close(bio_file, BIO_CLOSE);
        bio_chain = BIO_push(bio_file, bio_chain);

        bio_b64 = BIO_new(BIO_f_base64());
        if (!bio_b64) {
                printf("ERROR: BIO_new(f_base64)\n");
                return(2);
        }
        bio_chain = BIO_push(bio_b64, bio_chain);

        bio_cipher = BIO_new(BIO_f_cipher());
        if (!bio_cipher) {
                printf("ERROR: BIO_new()\n");
                return(2);
        }
        bio_chain = BIO_push(bio_cipher, bio_chain);

        bio_chain = BIO_find_type(bio_chain, BIO_TYPE_CIPHER);
        if (!bio_chain) {
                printf("ERROR: BIO_find_type()\n");
                return(2);
        }

        print_bio_chain(bio_chain);


        /* Turn on encryption */
        /*
        BIO_set_cipher(bio_chain, EVP_aes_256_cbc(), key, iv, 1);
        */
        BIO_set_cipher(bio_chain, EVP_bf_cbc(), key, iv, 1);


        /* Write */

        printf("data(%zd) = %s\n", strlen(data), data);

        remaining = strlen(data);
        while (remaining > 0) {
                ret = BIO_write(bio_chain, data, remaining);

                if (ret <= 0) {
                        if (BIO_should_retry(bio_chain)) {
                                printf("write delay\n");

                                sleep(1);
                                continue;
                        } else {
                                printf("BIO_write() error (don't retry)\n");

                                break;
                        }
                }

                remaining -= ret;

                printf("wrote: %d\n", ret);
        }

        if (BIO_get_cipher_status(bio_chain) == 0) {
                printf("ERROR: failed to encrypt\n");
        } else {
                printf("Encrypt OK\n");
        }

        /* Flush */
        do {
                if (BIO_flush(bio_chain) == 1) {
                        printf("flushed bio_chain\n");
                } else {
                        if (BIO_should_retry(bio_chain)) {
                                printf("flush delay\n");

                                sleep(1);
                                continue;
                        } else {
                                printf("BIO_flush() error (don't retry)\n");

                                break;
                        }
                }
                printf("pending write bytes from bio_chain: %d\n", BIO_wpending(bio_chain));
                printf("pending bytes from bio_chain: %d\n", BIO_pending(bio_chain));
        } while (BIO_wpending(bio_chain) > 0);

        printf("db_file size -> %d\n", BIO_tell(bio_chain));
        remaining = BIO_tell(bio_chain);


        /* Reset */

        if (BIO_seek(bio_chain, 0) != 0) {
                printf("BIO_seek() error\n");
                return(2);
        }

        /* Turn on decryption */
        /*
        BIO_set_cipher(bio_chain, EVP_aes_256_cbc(), key, iv, 0);
        */
        BIO_set_cipher(bio_chain, EVP_bf_cbc(), key, iv, 1);


        /* Read */

        pos = 0;
        read = malloc(remaining + 1);
        do {
                ret = BIO_read(bio_chain, read + pos, remaining - pos);
                printf("read: %d\n", ret);

                pos += ret;
        } while (BIO_eof(bio_chain) < 1);

        if (BIO_get_cipher_status(bio_chain) == 0) {
                printf("ERROR: failed to decrypt\n");
        } else {
                printf("Decrypt OK\n");

                read[pos] = '\0';
                printf("data(%zd) = %s\n", strlen(read), read);
        }

}


void
print_bio_chain(BIO *bio_chain)
{
        BIO     *bio_tmp = NULL;


        printf("bio_chain: ");
        bio_tmp = bio_chain;
        while (bio_tmp) {
                switch (BIO_method_type(bio_tmp)) {
                        case (2|0x0400):
                                printf("BIO_TYPE_FILE");
                        break;
                        case (10|0x0200):
                                printf("BIO_TYPE_CIPHER");
                        break;
                        case (11|0x0200):
                                printf("BIO_TYPE_BASE64");
                        break;
                        default:
                                printf("%04X", BIO_method_type(bio_tmp));
                        break;
                }
                bio_tmp = BIO_next(bio_tmp);
                if (!bio_tmp)
                        break;
                printf("->");
        }
        puts("");
} /* print_bio_chain */
```

[test.c.txt](https://github.com/openssl/openssl/files/9969331/test.c.txt)

**Comment by beldmit** [comment]
As BF is moved to the legacy provider, you have to load it to make the algorithm available.

**Comment by mattcaswell** [comment]
You don't check the return value form `BIO_set_cipher`. Almost certainly it is failing due to the legacy provider not having been loaded. I guess that is leaving the cipher BIO in some indeterminate state. You need to load the default and legacy providers explicitly (either via config or in code).

**Comment by levaidaniel** [comment]
> As BF is moved to the legacy provider, you have to load it to make the algorithm available.

Indeed, using
```
OSSL_PROVIDER_load(NULL, "legacy")
```
loads BF and makes it working again.

Thank you!

**Comment by levaidaniel** [comment]
> You don't check the return value form `BIO_set_cipher`. Almost certainly it is failing due to the legacy provider not having been loaded. I guess that is leaving the cipher BIO in some indeterminate state. You need to load the default and legacy providers explicitly (either via config or in code).

You're right; to my excuse, BIO_set_cipher() didn't return anything in earlier versions (e.g. 1.1.1), so no check was implemented.
Loading the _legacy_ provider makes BF available again, thanks!

**Comment by mattcaswell** [comment]
> BIO_set_cipher() didn't return anything in earlier versions (e.g. 1.1.1)

Actually this appears to be a documentation error in the 1.1.1 docs. In fact `BIO_set_cipher` has returned an error since 1.1.0 (1.0.2 and earlier did indeed not return an error). There's a CHANGES entry about it for the 1.1.0 release:

````
  *) Extensive audit of libcrypto with DEBUG_UNUSED. Fix many cases where
     return value is ignored. NB. The functions RAND_add(), RAND_seed(),
     BIO_set_cipher() and some obscure PEM functions were changed so they
     can now return an error. The RAND changes required a change to the
     RAND_METHOD structure.
     [Steve Henson]
````

It seems the 1.1.x docs did not get updated to reflect this change - although they have subsequently been fixed for 3.0
